Bump node version, add tests and code coverage

garrytrinder opened the pull request:
Merging this PR will

- Bump action to `node@16`
  - Causes major version bump to `v4.0.0`(see https://github.blog/changelog/2022-05-20-actions-can-now-run-in-a-node-js-16-runtime/)
- Update contributors
- Add mocha unit tests
- Add c8 code coverage
- Add support for Mocha Test Explorer VS Code extension
- Add test workflows
- Add PR check workflow
- Update dependencies
- Fix dependency versions
- Update contributors
- Remove `.DS_Store` file
  - Add to .gitignore 
- Update readme
  - Update compatibility matrix
  - Remove release notes
- Remove tslint
- Add eslint rules
   - Two space Indenting
   - Single quote rule, quote literals allowed
- Update npm scripts

garrytrinder committed:
Bump node version, add tests and code coverage

Jwaegebaert commented on .c8rc.json:
Let's use a spacing indent here of 2 instead of 4

Jwaegebaert commented on .mocharc.json:
Let's use a spacing indent here of 2 instead of 4

Jwaegebaert commented on README.md:
```diff
@@ -100,30 +101,5 @@ jobs:
 #### Self-hosted runners
 If self-hosted runners are used for running the workflow, then please make sure that they have `PowerShell` or `bash` installed on them. 
-
-## Release notes
-
-### v3.0.1
-
-- Fixes compatibility bug with CLI for Microsoft 365 v6.0.0
-
-### v3.0.0
-
-- Ensured compatibility with CLI for Microsoft 365 v6.0.0
-
-### v2.0.2
-
-- Fixes bug where app ID was invalid
-
-### v2.0.0
-
-- Renames action to 'CLI for Microsoft 365' 
-
-### v1.0.1
-- Fixed 'skipFeatureDeployment not included in spo app deploy command for Windows' solving #4
-
-### v1.0.0
-- Added inital 'CLI for Microsoft 365 deploy' GitHub action solving #2
```
Why would we remove the release notes?

garrytrinder replied:
I removed them from `action-cli-login` repo readme to make the readme more concise and less cluttered. As we already have the releases with the notes on the releases area users can view the changes over time that way.

Jwaegebaert commented on action.yml:
```diff
@@ -1,23 +1,30 @@
-name: 'CLI for Microsoft 365 Deploy App'
-description: 'Deploy an app using CLI for Microsoft 365'
+name: "CLI for Microsoft 365 Deploy App"
```
I would prefer to use single quotes instead of double quotes as we also use single quotes in the CLI 😄
But this is a personal preference of course.

Jwaegebaert commented on package.json:
```diff
@@ -1,23 +1,51 @@
+    "test:cov": "c8 npm run test"
   },
-  "author": "Tatti, Anoop <anoop@live.co.uk>",
+  "keywords": [
+    "office 365",
+    "microsoft 365",
+    "sharepoint framework",
+    "o365",
+    "m365",
+    "spfx",
+    "sharepoint online",
+    "sharepoint",
+    "microsoft teams",
+    "microsoft graph",
+    "microsoft flow",
+    "azure active directory",
+    "azure ad",
+    "azure",
+    "microsoft"
+  ],
```
Maybe a few additional keywords regarding power platform could be handy here

garrytrinder replied:
@Jwaegebaert good catch, these are the same keywords we have in the main CLI repo, so we should consider also adding power platform there as well.

garrytrinder replied:
I think these keywords need a review. Let's leave them as they are for now and work on a new set of keywords that we can apply across all of our repos.

Jwaegebaert replied:
Good idea, I'll create an issue in the repo of the CLI with a suggestion and we can discuss it there.

garrytrinder replied:
Thank you!

Jwaegebaert commented on package.json:
```diff
@@ -32,48 +60,62 @@
+    },
+    {
+      "name": "Jasey Waegebaert",
+      "email": "jaseyw@gmigroup.be",
```
```suggestion
      "email": "38426621+Jwaegebaert@users.noreply.github.com",
```

Jwaegebaert commented on src/commands.spec.ts:
Let's use a spacing indent here of 2 instead of 4

Jwaegebaert commented on src/commands.spec.ts:
```diff
@@ -0,0 +1,113 @@
+import { getAddCommand, getDeployCommand, } from './commands';
+import { Options, Scope } from './validate';
+
+describe('commands', () => {
+
+    describe('getAddCommand', () => {
+
+        it('returns correct command with options (defaults)', () => {
```
```suggestion
describe('commands', () => {
    describe('getAddCommand', () => {
        it('returns correct command with options (defaults)', () => {
```

Jwaegebaert commented on src/commands.spec.ts:
```diff
@@ -0,0 +1,113 @@
+
+    });
+
+    describe('getDeployCommand', () => {
+
+        it('returns correct command with options (defaults)', () => {
```
```suggestion
    describe('getDeployCommand', () => {
        it('returns correct command with options (defaults)', () => {
```

Jwaegebaert commented on src/commands.ts:
Let's use a spacing indent here of 2 instead of 4

Jwaegebaert commented on src/commands.ts:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    if (options.OVERWRITE) {
+        command += " --overwrite";
```
I would prefer to use single quotes here so we are consistent with the test file.

```suggestion
        command += ' --overwrite';
```

Jwaegebaert commented on src/commands.ts:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    if (options.SKIP_FEATURE_DEPLOYMENT) {
+        command += " --skipFeatureDeployment";
```
```suggestion
        command += ' --skipFeatureDeployment';
```

Jwaegebaert commented on src/commands.spec.ts:
```diff
@@ -0,0 +1,113 @@
+                OVERWRITE: false
+            };
+            const command = getAddCommand(options);
+            assert.equal(command, `spo app add --filePath solution.sppkg`);
```
```suggestion
            assert.equal(command, 'spo app add --filePath solution.sppkg');
```

Jwaegebaert commented on src/commands.spec.ts:
```diff
@@ -0,0 +1,113 @@
+                OVERWRITE: false
+            };
+            const command = getAddCommand(options);
+            assert.equal(command, `spo app add --filePath solution.sppkg --appCatalogScope sitecollection --appCatalogUrl https://contoso.sharepoint.com/sites/site1`);
```
```suggestion
            assert.equal(command, 'spo app add --filePath solution.sppkg --appCatalogScope sitecollection --appCatalogUrl https://contoso.sharepoint.com/sites/site1');
```

Jwaegebaert commented on src/commands.spec.ts:
```diff
@@ -0,0 +1,113 @@
+                OVERWRITE: true
+            };
+            const command = getAddCommand(options);
+            assert.equal(command, `spo app add --filePath solution.sppkg --overwrite`);
```
```suggestion
            assert.equal(command, 'spo app add --filePath solution.sppkg --overwrite');
```

Jwaegebaert commented on src/commands.spec.ts:
```diff
@@ -0,0 +1,113 @@
+                OVERWRITE: true
+            };
+            const command = getAddCommand(options);
+            assert.equal(command, `spo app add --filePath solution.sppkg --appCatalogScope sitecollection --appCatalogUrl https://contoso.sharepoint.com/sites/site1 --overwrite`);
```
```suggestion
            assert.equal(command, 'spo app add --filePath solution.sppkg --appCatalogScope sitecollection --appCatalogUrl https://contoso.sharepoint.com/sites/site1 --overwrite');
```

Jwaegebaert commented on src/commands.spec.ts:
```diff
@@ -0,0 +1,113 @@
+                OVERWRITE: false
+            };
+            const command = getDeployCommand(options, appId);
+            assert.equal(command, `spo app deploy --id 00000000-0000-0000-0000-000000000000`);
```
```suggestion
            assert.equal(command, 'spo app deploy --id 00000000-0000-0000-0000-000000000000');
```

Jwaegebaert commented on src/commands.spec.ts:
```diff
@@ -0,0 +1,113 @@
+                OVERWRITE: false
+            };
+            const command = getDeployCommand(options, appId);
+            assert.equal(command, `spo app deploy --id 00000000-0000-0000-0000-000000000000 --appCatalogScope sitecollection --appCatalogUrl https://contoso.sharepoint.com/sites/site1`);
```
```suggestion
            assert.equal(command, 'spo app deploy --id 00000000-0000-0000-0000-000000000000 --appCatalogScope sitecollection --appCatalogUrl https://contoso.sharepoint.com/sites/site1');
```

Jwaegebaert commented on src/commands.spec.ts:
```diff
@@ -0,0 +1,113 @@
+                OVERWRITE: false
+            };
+            const command = getDeployCommand(options, appId);
+            assert.equal(command, `spo app deploy --id 00000000-0000-0000-0000-000000000000 --skipFeatureDeployment`);
```
```suggestion
            assert.equal(command, 'spo app deploy --id 00000000-0000-0000-0000-000000000000 --skipFeatureDeployment');
```

Jwaegebaert commented on src/commands.spec.ts:
```diff
@@ -0,0 +1,113 @@
+                OVERWRITE: true
+            };
+            const command = getDeployCommand(options, appId);
+            assert.equal(command, `spo app deploy --id 00000000-0000-0000-0000-000000000000 --appCatalogScope sitecollection --appCatalogUrl https://contoso.sharepoint.com/sites/site1 --skipFeatureDeployment`);
```
```suggestion
            assert.equal(command, 'spo app deploy --id 00000000-0000-0000-0000-000000000000 --appCatalogScope sitecollection --appCatalogUrl https://contoso.sharepoint.com/sites/site1 --skipFeatureDeployment');
```

Jwaegebaert commented on src/constants.ts:
Let's use a spacing indent here of 2 instead of 4 and single quotes as well

Jwaegebaert commented on src/main.ts:
Let's use a spacing indent here of 2 instead of 4

Jwaegebaert commented on src/models.ts:
Let's use a spacing indent here of 2 instead of 4

Jwaegebaert commented on src/utils.spec.ts:
Let's use a spacing indent here of 2 instead of 4

Jwaegebaert commented on src/utils.spec.ts:
```diff
@@ -0,0 +1,128 @@
+
+    describe('isNullOrEmpty', () => {
+
+        it('returns true when empty string is passed', () => {
```
```suggestion
describe('utils', () => {
    describe('isNullOrEmpty', () => {
        it('returns true when empty string is passed', () => {
```

Jwaegebaert commented on src/utils.spec.ts:
```diff
@@ -0,0 +1,128 @@
+        });
+    });
+
+    describe('isStringBoolean', () => {
+
+        it('returns true when string is "true"', () => {
```
```suggestion
    describe('isStringBoolean', () => {
        it('returns true when string is "true"', () => {
```

Jwaegebaert commented on src/utils.spec.ts:
```diff
@@ -0,0 +1,128 @@
+
+    describe('getOptions', () => {
+
+        it('throws error when no options are passed', () => {
```
```suggestion
    describe('getOptions', () => {
        it('throws error when no options are passed', () => {
```

Jwaegebaert commented on src/utils.spec.ts:
```diff
@@ -0,0 +1,128 @@
+        });
+    });
+
+    describe('execCommand', () => {
+
+        it('returns correct object', async () => {
```
```suggestion
    describe('execCommand', () => {
        it('returns correct object', async () => {
```

Jwaegebaert commented on src/utils.ts:
Let's use a spacing indent here of 2 instead of 4

Jwaegebaert commented on src/validate.spec.ts:
Let's use a spacing indent here of 2 instead of 4

Jwaegebaert commented on src/validate.spec.ts:
```diff
@@ -0,0 +1,60 @@
+import * as assert from 'assert';
+import { Options, Scope, validate } from './validate';
+
+describe('validate', () => {
+
+    it('throws error if APP_FILE_PATH is not passed ', () => {
```
```suggestion
describe('validate', () => {
    it('throws error if APP_FILE_PATH is not passed ', () => {
```

Jwaegebaert commented on src/validate.spec.ts:
```diff
@@ -0,0 +1,60 @@
+            SKIP_FEATURE_DEPLOYMENT: false,
+            OVERWRITE: false
+        };
+        assert.throws(() => validate(options), Error, `SCOPE must be either 'tenant' or Scope.SiteCollection.`);
```
```suggestion
        assert.throws(() => validate(options), Error, 'SCOPE must be either 'tenant' or Scope.SiteCollection.');
```

Jwaegebaert commented on src/validate.spec.ts:
```diff
@@ -0,0 +1,60 @@
+            SKIP_FEATURE_DEPLOYMENT: false,
+            OVERWRITE: false
+        };
+        assert.throws(() => validate(options), Error, `SITE_COLLECTION_URL is required if SCOPE is set to Scope.SiteCollection.`);
```
```suggestion
        assert.throws(() => validate(options), Error, 'SITE_COLLECTION_URL is required if SCOPE is set to Scope.SiteCollection.');
```

Jwaegebaert commented on src/validate.ts:
Let's use a spacing indent here of 2 instead of 4

Jwaegebaert reviewed (changes requested):
Awesome work with this PR @garrytrinder. I'm no master of GitHub actions but at first glance, they look all right. I've left a few suggestions but those are mostly cosmetic 😉

garrytrinder commented:
Thanks for the review comments @Jwaegebaert 👍

I've addressed the indenting and quoting comments by applying eslint rules so we fix the style going forwards.

I've taken the approach of using no quotes > single quotes > double quotes for the `action.yml` file. This seems to be the best approach based on a little research, (see https://stackoverflow.com/a/69850618).

I've also updated some of the npm scripts to bring them inline with the main CLI repo so that we have a more consistent developer experience.

Jwaegebaert commented on .github/workflows/check_pr.yml:
```diff
@@ -0,0 +1,93 @@
+
+    steps:
+      - uses: actions/checkout@v3
+      - name: Use Node.js ${{ matrix.node }}
+        uses: actions/setup-node@v3
+        with:
+          node-version: ${{ matrix.node }}
+          registry-url: "https://registry.npmjs.org"
+      - name: Cache node modules
+        id: cache
+        uses: actions/cache@v3
+        with:
+          path: |
+            **/node_modules
+          key: node_modules-${{ matrix.os }}-${{ matrix.node }}-${{ hashFiles('**/npm-shrinkwrap.json') }}
+      - name: Restore dependencies
+        if: steps.cache.outputs.cache-hit != 'true'
+        run: npm ci
+      - name: Build
+        run: npm run build
+      - name: Compress output (non-Windows)
+        if: matrix.os != 'windows-latest'
+        run: tar -cvf build.tar --exclude node_modules ./
+      - name: Compress output (Windows)
+        if: matrix.os == 'windows-latest'
+        run: 7z a -ttar -xr!node_modules -r build.tar .
+      - name: Upload build artifact
+        uses: actions/upload-artifact@v3
+        with:
+          name: build-${{ matrix.os }}-${{ matrix.node }}
+          path: build.tar
```
In `.github/workflows/test.yml` we have an empty line before a new task which looks quite neat. So a suggestion would be to apply it here as well 😄 

```suggestion
      - name: Use Node.js ${{ matrix.node }}
        uses: actions/setup-node@v3
        with:
          node-version: ${{ matrix.node }}
          registry-url: "https://registry.npmjs.org"
          
      - name: Cache node modules
        id: cache
        uses: actions/cache@v3
        with:
          path: |
            **/node_modules
          key: node_modules-${{ matrix.os }}-${{ matrix.node }}-${{ hashFiles('**/npm-shrinkwrap.json') }}
          
      - name: Restore dependencies
        if: steps.cache.outputs.cache-hit != 'true'
        run: npm ci
        
      - name: Build
        run: npm run build
        
      - name: Compress output (non-Windows)
        if: matrix.os != 'windows-latest'
        run: tar -cvf build.tar --exclude node_modules ./
        
      - name: Compress output (Windows)
        if: matrix.os == 'windows-latest'
        run: 7z a -ttar -xr!node_modules -r build.tar .
        
      - name: Upload build artifact
        uses: actions/upload-artifact@v3
        with:
          name: build-${{ matrix.os }}-${{ matrix.node }}
          path: build.tar
```

Jwaegebaert commented on .github/workflows/check_pr.yml:
```diff
@@ -0,0 +1,93 @@
+    steps:
+      - name: Configure pagefile
+        if: matrix.os == 'windows-latest'
+        uses: al-cheb/configure-pagefile-action@v1.3
+        with:
+          minimum-size: 16GB
+          maximum-size: 16GB
+          disk-root: "C:"
+      - uses: actions/download-artifact@v3
+        with:
+          name: build-${{ matrix.os }}-${{ matrix.nodeBuild }}
+      - name: Unpack build artifact (non-Windows)
+        if: matrix.os != 'windows-latest'
+        run: tar -xvf build.tar && rm build.tar
+      - name: Unpack build artifact (Windows)
+        if: matrix.os == 'windows-latest'
+        run: 7z x build.tar && del build.tar
+      - name: Use Node.js ${{ matrix.nodeRun }}
+        uses: actions/setup-node@v3
+        with:
+          node-version: ${{ matrix.nodeRun }}
+          registry-url: "https://registry.npmjs.org"
+      - name: Cache node modules
+        id: cache
+        uses: actions/cache@v3
+        with:
+          path: |
+            **/node_modules
+          key: node_modules-${{ matrix.os }}-${{ matrix.nodeBuild }}-${{ hashFiles('**/npm-shrinkwrap.json') }}
+      - name: Restore dependencies
+        if: steps.cache.outputs.cache-hit != 'true'
+        run: npm ci
+      - name: Test without coverage
+        run: npm test
+      - name: Test with coverage
+        run: npm run test:cov
+      - uses: actions/upload-artifact@v3
+        with:
+          name: coverage-${{ matrix.os }}-${{ matrix.nodeRun }}
+          path: coverage.tar
```
Let's add an empty line before every task.

```suggestion
      - name: Configure pagefile
        if: matrix.os == 'windows-latest'
        uses: al-cheb/configure-pagefile-action@v1.3
        with:
          minimum-size: 16GB
          maximum-size: 16GB
          disk-root: "C:"
          
      - uses: actions/download-artifact@v3
        with:
          name: build-${{ matrix.os }}-${{ matrix.nodeBuild }}
          
      - name: Unpack build artifact (non-Windows)
        if: matrix.os != 'windows-latest'
        run: tar -xvf build.tar && rm build.tar
        
      - name: Unpack build artifact (Windows)
        if: matrix.os == 'windows-latest'
        run: 7z x build.tar && del build.tar
        
      - name: Use Node.js ${{ matrix.nodeRun }}
        uses: actions/setup-node@v3
        with:
          node-version: ${{ matrix.nodeRun }}
          registry-url: "https://registry.npmjs.org"
          
      - name: Cache node modules
        id: cache
        uses: actions/cache@v3
        with:
          path: |
            **/node_modules
          key: node_modules-${{ matrix.os }}-${{ matrix.nodeBuild }}-${{ hashFiles('**/npm-shrinkwrap.json') }}
          
      - name: Restore dependencies
        if: steps.cache.outputs.cache-hit != 'true'
        run: npm ci
        
      - name: Test without coverage
        run: npm test
        
      - name: Test with coverage
        run: npm run test:cov
        
      - uses: actions/upload-artifact@v3
        with:
          name: coverage-${{ matrix.os }}-${{ matrix.nodeRun }}
          path: coverage.tar
```

Jwaegebaert commented on .c8rc.json:
```diff
@@ -0,0 +1,16 @@
+{
+    "exclude": [
```
Here we still have a spacing of 4 instead of 2

Jwaegebaert commented on .eslintrc.json:
```diff
@@ -0,0 +1,35 @@
+{
+    "root": true,
```
Here we also have a spacing of 4 instead of 2

Jwaegebaert commented on src/models.ts:
```diff
@@ -0,0 +1,24 @@
+export interface SpoApp {
+    CheckInComment: string,
```
Here we also have a spacing of 4 instead of 2

Jwaegebaert commented on src/validate.ts:
```diff
@@ -0,0 +1,36 @@
+import { isNullOrEmpty } from './utils';
+
+export interface Options {
+    APP_FILE_PATH: string;
+    SCOPE: Scope;
+    SITE_COLLECTION_URL: string;
+    SKIP_FEATURE_DEPLOYMENT: boolean;
+    OVERWRITE: boolean;
+}
+
+export enum Scope {
+    Tenant = 'tenant',
+    SiteCollection = 'sitecollection'
+}
```
Here we also have a spacing of 4 instead of 2

```suggestion
export interface Options {
  APP_FILE_PATH: string;
  SCOPE: Scope;
  SITE_COLLECTION_URL: string;
  SKIP_FEATURE_DEPLOYMENT: boolean;
  OVERWRITE: boolean;
}

export enum Scope {
  Tenant = 'tenant',
  SiteCollection = 'sitecollection'
}
```

Jwaegebaert commented on src/validate.spec.ts:
```diff
@@ -0,0 +1,60 @@
+      SKIP_FEATURE_DEPLOYMENT: false,
+      OVERWRITE: false
+    };
+    assert.throws(() => validate(options), Error, `SITE_COLLECTION_URL is required if SCOPE is set to Scope.SiteCollection.`);
```
```suggestion
    assert.throws(() => validate(options), Error, 'SITE_COLLECTION_URL is required if SCOPE is set to Scope.SiteCollection.');
```

Jwaegebaert commented on .mocharc.json:
```diff
@@ -0,0 +1,11 @@
+{
+    "extension": [
```
Here we also have a spacing of 4 instead of 2

Jwaegebaert reviewed (changes requested):
Awesome idea to add the eslint rules. I noticed a few files still have a spacing of 4 instead of 2, but I commented on them. I also found another nice addition we could apply. Besides these, it looks good to go 😄

Jwaegebaert reviewed (approved):
Great job on this one, @garrytrinder! I don't have anything else to add to this PR, so it's a go from me 😄

garrytrinder commented:
Thanks again @Jwaegebaert 👍 I managed to fix the eslint rules so that we can now detect errors in the interfaces and enums which were previously being ignored. Thanks again for your attention to detail ❤️
